### PR TITLE
Backlog/definition object/implement definition object pipeline

### PR DESCRIPTION
**Pull request by lluisFtrack**
Resolves : 

* CLICKUP- https://app.clickup.com/t/3jv10m8

- [ ] I have added automatic tests where applicable.
- [ ] The PR contains a description of what has been changed.
- [ ] The description contains manual test instructions.
- [ ] The PR contains update to the release notes.
- [ ] The PR contains update to the documentation.

This PR has been tested on :

- [ ] Windows.
- [ ] MacOs.
- [ ] Linux.

## Changes
DefinitionObject is implemented and used by the client in the pipelien repository. 
Relevant notes:
- Host connection converts all the definitions to definition objects so clients are always dealing with definition objects.
- Swap arguments of the `change_definition` method and schema now is optional
- Optional arguments on the `run_definition` method
- DefinitionObject and DefinitionList has `get_all` and `get_first` methods. (`get_categories`, `get_by_name` and `get_by_type` are removed)
- Client has the property definitions, so we can query all the definitions using client.definitions instead of client.host_connection.definitions

## Test
Snipped code attached with an example publishing in 100% standalone mode and with an example publishing from maya python console.
( PLEASE BE AWARE THAT YOU WILL HAVE TO CHANGE EXAMPLE FILES EXTENSION TO .py, NOT UPLOADED THIS WAY BECAUSE GITHUB DOESN?T SUPPORT .py EXTENSION)

[publish_from_console_test.py.txt](https://github.com/ftrackhq/ftrack-connect-pipeline/files/9775357/publish_from_console_test.py.txt)

[DCC_publish_standalone.py.txt](https://github.com/ftrackhq/ftrack-connect-pipeline/files/9775361/DCC_publish_standalone.py.txt)

### CODE REVIEW

File: source/ftrack_connect_pipeline/client/asset_manager.py [hunk 1]
```diff
@@ -37,7 +37,7 @@ def on_host_changed(self, host_connection):
         definitions = self.host_connection.definitions.get(schema_title)
         if len(definitions) > 0:
             # Only one definition for now, we don't have a definition schema on the AM
-            self.change_definition(schema, definitions[0])
+            self.change_definition(definitions[0], schema)
```
Comment (hdd) [inline source/ftrack_connect_pipeline/client/asset_manager.py:40]: @lluisFtrack how about **set_definition** ? 
seems more suitable then **change_definition** (as in change from ? ... )

Comment (henriknorin-ftrack) [reply source/ftrack_connect_pipeline/client/asset_manager.py:40]: Or just have a simple setter?

Comment (henriknorin-ftrack) [reply source/ftrack_connect_pipeline/client/asset_manager.py:40]: Or just have a simple setter?

File: source/ftrack_connect_pipeline/definition/definition_object.py [hunk 2]
```diff
@@ -188,15 +184,39 @@ def __init__(self, options):
...
+                    issubclass(type(item), DefinitionList) or
+                    issubclass(type(item), DefinitionObject)
+            ):
+                result = item.get_all(first=first, **kwargs)
+                # Return the first value that matches if first is true
+                if result and first:
+                    return result
+                elif result:
+                    results.extend(result)
+        return results
```
[the remark this answers is not in the document]
Comment (lluisFtrack) [reply source/ftrack_connect_pipeline/definition/definition_object.py:205]: The expected result is an empty list which is what is returning. Do we really want to return None?

**Comment by henriknorin-ftrack** [comment]
Got an validation error when running a definition object, could fix it by converting definition to dictionary before validation:

```
/ Error: ERROR - MainThread - 2022-10-17 13:35:57,534 - ftrack_connect_pipeline.host.MayaHost - Can't validate the data DefinitionObject(
..)  is not of type 'object'
// 
// Failed validating 'type' in schema:
//     {'additionalProperties': False,
.. On instance:
//     DefinitionObject({
...


```

File: source/ftrack_connect_pipeline/definition/definition_object.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lluisFtrack) [reply source/ftrack_connect_pipeline/definition/definition_object.py:205]: True that. I'll change it

[the remark this answers is not in the document]
Comment (lluisFtrack) [reply source/ftrack_connect_pipeline/definition/definition_object.py:205]: Changed, get_first is returning None

**Comment by lluisFtrack** [comment]
> Got an validation error when running a definition object, could fix it by converting definition to dictionary before validation:
> 
> ```
> / Error: ERROR - MainThread - 2022-10-17 13:35:57,534 - ftrack_connect_pipeline.host.MayaHost - Can't validate the data DefinitionObject(
> ..)  is not of type 'object'
> // 
> // Failed validating 'type' in schema:
> //     {'additionalProperties': False,
> .. On instance:
> //     DefinitionObject({
> ...
> ```

Done

**Comment by lluisFtrack** [comment]
> Also got an exception on running a definition object:
> 
> ```
> Traceback (most recent call last):
>  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_api/event/hub.py", line 782, in _handle
>   response = subscriber.callback(event)
>  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-maya-1.0.2/dependencies/ftrack_connect_pipeline_maya/host/__init__.py", line 37, in run
>   runnerResult = super(MayaHost, self).run(event)
>  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_connect_pipeline/host/__init__.py", line 152, in run
>   runner_result = engine_runner.run_definition(data)
>  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_connect_pipeline/host/engine/__init__.py", line 698, in run_definition
>   components_output[i]['result'].remove(
> ValueError: list.remove(x): x not in list
> ```
> 
> Otherwise my tests turned out well and are ready to approve PR as soon as the issues have been addressed.

@henriknorin-ftrack can you please add the steps you followed to reproduce this error? So we can make sure that is solved. Thanks.

**Comment by henriknorin-ftrack** [comment]
> > Also got an exception on running a definition object:
> > ```
> > Traceback (most recent call last):
> >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_api/event/hub.py", line 782, in _handle
> >   response = subscriber.callback(event)
> >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-maya-1.0.2/dependencies/ftrack_connect_pipeline_maya/host/__init__.py", line 37, in run
> >   runnerResult = super(MayaHost, self).run(event)
> >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_connect_pipeline/host/__init__.py", line 152, in run
> >   runner_result = engine_runner.run_definition(data)
> >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_connect_pipeline/host/engine/__init__.py", line 698, in run_definition
> >   components_output[i]['result'].remove(
> > ValueError: list.remove(x): x not in list
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Otherwise my tests turned out well and are ready to approve PR as soon as the issues have been addressed.
> 
> @henriknorin-ftrack can you please add the steps you followed to reproduce this error? So we can make sure that is solved. Thanks.

You can try publishing with my experimental branch, that one uses your definition object fully out: backlog/definitionobject/experimental

**Comment by lluisFtrack** [comment]
> > > Also got an exception on running a definition object:
> > > ```
> > > Traceback (most recent call last):
> > >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_api/event/hub.py", line 782, in _handle
> > >   response = subscriber.callback(event)
> > >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-maya-1.0.2/dependencies/ftrack_connect_pipeline_maya/host/__init__.py", line 37, in run
> > >   runnerResult = super(MayaHost, self).run(event)
> > >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_connect_pipeline/host/__init__.py", line 152, in run
> > >   runner_result = engine_runner.run_definition(data)
> > >  File "/Users/henriknorin/Library/Application Support/ftrack-connect-plugins/ftrack-connect-pipeline-1.0.3/dependencies/ftrack_connect_pipeline/host/engine/__init__.py", line 698, in run_definition
> > >   components_output[i]['result'].remove(
> > > ValueError: list.remove(x): x not in list
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Otherwise my tests turned out well and are ready to approve PR as soon as the issues have been addressed.
> > 
> > 
Screenshot 2022-10-24 at 11 22 09

> > @henriknorin-ftrack can you please add the steps you followed to reproduce this error? So we can make sure that is solved. Thanks.
> 
> You can try publishing with my experimental QT branch, that one uses your definition object fully out: backlog/definitionobject/experimental

Hi @henriknorin-ftrack can't reproduce the error, and seems that the qt experimental branch has something odd (see the attached image), please can you make sure that this is still happening to you? ( I updated somethings on this PR so make sure you use the lates version)
Also, I don't think this error is related as it seems to be complaining on the host which is not been touched. Happy to review it with you if you like.

**Review by henriknorin-ftrack — CHANGES_REQUESTED** [review]
Please run Black on it

**Review by henriknorin-ftrack — APPROVED** [review]

**Pull request merged**